### PR TITLE
MERC 2253 use multiple latest blocks to improve consensus v 0 2

### PR DESCRIPTION
**Pull request by samsondav**

### CODE REVIEW

File: pkg/reportingplugins/mercury/v1/mercury.go [hunk 1]
```diff
@@ -248,7 +249,40 @@ func parseAttributedObservation(ao ocrtypes.AttributedObservation) (PAO, error)
...
 
-	if obs.CurrentBlockValid {
+	if len(obs.LatestBlocks) > 0 {
+		for _, b := range obs.LatestBlocks {
+			pao.LatestBlocks = append(pao.LatestBlocks, NewBlock(b.Num, b.Hash, b.Ts))
+
+			// Ignore observation if it has duplicate blocks by number or hash
+			// for security to avoid the case where one node can "throw" block
+			// numbers by including a bunch of duplicates
+			// TODO: test this
```
Comment (samsondav) [inline pkg/reportingplugins/mercury/v1/mercury.go:259]: dont forget to add this test

File: pkg/reportingplugins/mercury/v1/mercury.go [hunk 2]
```diff
@@ -248,7 +249,40 @@ func parseAttributedObservation(ao ocrtypes.AttributedObservation) (PAO, error)
 		pao.PricesValid = true
 	}
 
-	if obs.CurrentBlockValid {
+	if len(obs.LatestBlocks) > 0 {
```
Comment (kaleofduty) [inline pkg/reportingplugins/mercury/v1/mercury.go:252]: probably also worth discarding an observation if it has _too many_ latest blocks

File: pkg/reportingplugins/mercury/v1/types.go [hunk 3]
```diff
@@ -9,14 +10,59 @@ import (
...
+	if b.Num == b2.Num && b.Ts == b2.Ts {
+		// tie-break on hash, all else being equal
+		return b.Hash > b2.Hash
+	} else if b.Num == b2.Num {
+		// if block number is equal and timestamps differ, take the oldest timestamp
+		return b.Ts < b2.Ts
+	} else {
+		// if block number is different, take the lower block number
+		return b.Num < b2.Num
+	}
```
Comment (jmank88) [inline pkg/reportingplugins/mercury/v1/types.go:42]: The linter usually wants this form:
```suggestion
	}
	// if block number is different, take the lower block number
	return b.Num < b2.Num
```

Comment (samsondav) [reply pkg/reportingplugins/mercury/v1/types.go:42]: It doesn't seem to have complained about this one though... should I change it?

Comment (jmank88) [reply pkg/reportingplugins/mercury/v1/types.go:42]: Could leave it to see if it fails on develop as a test. It has been malfunctioning a lot lately

**Review by jmank88 — APPROVED** [review]

**Pull request merged**